### ISSUE TITLE
CloudSQL: support for Private networking

### ISSUE THREAD

**Issue by rolandkool** [terraform-providers/terraform-provider-google#2127]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment. If the issue is assigned to the "modular-magician" user, it is either in the process of being autogenerated, or is planned to be autogenerated soon. If the issue is assigned to a user, that user is claiming responsibility for the issue. If the issue is assigned to "hashibot", a community member has claimed the issue already.

### Description

Google recently announced that CloudSQL private networking moved to Beta: https://cloud.google.com/blog/products/databases/introducing-private-networking-connection-for-cloud-sql
It would be nice if the Google provider for terraform would add support for this feature as well.

### New or Affected Resource(s)

* google_sql_database_instance

**Comment by Stono** [comment]
Hey!
It'd be great if this was added, we kinda need it :-)

**Comment by gburiola** [comment]
@sereeth / @Stono what you can do in the meantime is enable private networking using `gcloud` and `local-exec`.

The gcloud commands are:

```
gcloud beta compute addresses create google-managed-services-default \
    --description='Peering range reserved for Google' --global \
    --network=default --purpose=VPC_PEERING --prefix-length=16

gcloud alpha services vpc-peerings connect --network=default \
    --ranges=google-managed-services-default --service=servicenetworking.googleapis.com
    
gcloud beta sql instances patch cloudsql-instance --network=default
```

```
resource "null_resource" "cloudsql-private-ip" {
  provisioner "local-exec" {
    "command" = "gcloud beta compute addresses create google-managed-services-default --description='Peering range reserved for Google' --global --network=default --purpose=VPC_PEERING --prefix-length=16 && gcloud alpha services vpc-peerings connect --network=default --ranges=google-managed-services-default --service=servicenetworking.googleapis.com && gcloud beta sql instances patch cloudsql-instance --network=default || true"
  }
}
```

The `|| true` at the end is because the last command takes more than 5 minutes to run and gcloud times out. There's no option on the gcloud command to specify the timeout so I reset the return code with `|| true`.
In my tests the last command always works despite the gcloud timeout

**Comment by sereeth** [comment]
@Stono I take it the network should be whatever network you want?

**Comment by craigatgoogle** [comment]
I'll be working on this. PR to follow :)

**Comment by craigatgoogle** [comment]
We're waiting on one final PR to merge for this to become available in the beta project: https://github.com/terraform-providers/terraform-provider-google-beta/pull/46

**Comment by glerma-sadasys** [comment]
Help!  approve PR ^  pretty please!

**Comment by glerma-sadasys** [comment]
The current beta for gcloud does not appear to support those arguments anymore.

```22:56 $ gcloud beta compute addresses create google-managed-services-default \
>     --description='Peering range reserved for Google' --global \
>     --network=agilone-vpc --purpose=VPC_PEERING --prefix-length=16
ERROR: (gcloud.beta.compute.addresses.create) unrecognized arguments:
  --network=agilone-vpc (did you mean '--network-tier'?)
  --purpose=VPC_PEERING (did you mean '--project'?)
  --prefix-length=16```

**Comment by robertpate** [comment]
I am running the local-exec right after creating the database and a user, and I am getting:
ERROR: (gcloud.beta.sql.instances.patch) HTTPError 409: Operation failed because another operation was already in progress.
Anybody seeing this?  Is there anything I can do to make the operation sequential?

**Comment by Chupaka** [comment]
Error on what? Be more descriptive, and create new issue :)

**Comment by glerma-sadasys** [comment]
> I am running the local-exec right after creating the database and a user, and I am getting:
> ERROR: (gcloud.beta.sql.instances.patch) HTTPError 409: Operation failed because another operation was already in progress.
> Anybody seeing this? Is there anything I can do to make the operation sequential?

I ran into this issue.  Essentially the cloud sql instance is still "in progress" when this happens.  I had to get around this by doing a little "depends on" trickery:  

```
resource "null_resource" "apply-cloudsql-private-ip" {
  provisioner "local-exec" {
    command = <<EOF
    gcloud beta compute addresses create ${var.peering_address_range_name}-${var.project_id} \
    --global \
    --addresses ${var.peering_cidr_range} \
    --prefix-length ${var.peering_cidr_prefix} \
    --description "Peering range for Google Services" \
    --project ${var.project_id} \
    --purpose VPC_PEERING  \
    --network ${var.network} \
    --quiet
EOF
  }
}

resource "null_resource" "destroy-cloudsql-private-ip" {
  provisioner "local-exec" {
    when = "destroy"

    command = <<EOF
    gcloud beta compute addresses delete ${var.peering_address_range_name}-${var.project_id} --global
EOF
  }
}

resource "null_resource" "vpc-to-services-peering" {
  provisioner "local-exec" {
    command = <<EOF
    gcloud beta services vpc-peerings connect \
    --service servicenetworking.googleapis.com \
    --network ${var.network} \
    --ranges "${var.peering_address_range_name}-${var.project_id}"  \
    --project=${var.project_id} \
    --quiet
EOF
  }

  depends_on = ["null_resource.apply-cloudsql-private-ip"]
}

resource "null_resource" "destroy-vpc-to-services-peering" {
  provisioner "local-exec" {
    when = "destroy"

    command = <<EOF
    gcloud compute networks peerings delete cloudsql-mysql-googleapis-com --network ${var.network} --quiet
    gcloud compute networks peerings delete servicenetworking-googleapis-com --network ${var.network} --quiet
EOF
  }
}

resource "null_resource" "patch_sql" {
  provisioner "local-exec" {
    command = <<EOF
    gcloud beta sql instances patch ${google_sql_database_instance.default.name}   --network ${var.network}  --quiet
    gcloud beta sql instances patch  ${google_sql_database_instance.default.name}  --no-assign-ip
EOF
  }

  // Do not change this order, otherwise can cause issues for you.
  depends_on = ["google_sql_database_instance.default", "google_sql_database.db", "google_sql_user.default", "null_resource.vpc-to-services-peering"]
}
```
It's not pretty, but it worked.  Hope that helps.

**Comment by craigatgoogle** [comment]
This indirect dependency is handled in the beta provider's implementation, see: https://github.com/terraform-providers/terraform-provider-google-beta/blob/2.0.0/website/docs/r/sql_database_instance.html.markdown

**Comment by adrianlop** [comment]
@craigatgoogle that's great news but unfortunately that `2.0.0` branch can't be used in terraform, afaik it has to be a Release so terraform can download it using `terraform init`, right?

**Comment by craigatgoogle** [comment]
I'll defer to @danawillow as to the best strategy for utilizing the beta provider.

**Comment by ipstatic** [comment]
Is there an ETA on that release? Or is there a way to run use this before the release?

**Comment by rileykarson** [comment]
You can [build from HEAD](https://github.com/terraform-providers/terraform-provider-google-beta#building-the-provider) on the `2.0.0` branch and [run it with a release copy of Terraform](https://github.com/terraform-providers/terraform-provider-google/blob/2.0.0/.github/CONTRIBUTING.md#running-a-local-copy-of-the-provider); we don't offer any guarantees of safety/stability so I wouldn't recommend doing so for anything other than testing purposes.

Our best ETA is that the `2.0.0` release will be roughly in line with Terraform 0.12's first beta.
